### ISSUE TITLE
Known issue (recoverable) in qpid-proton-j: uncorrelated channel

### ISSUE THREAD

**Issue by dgilling** [Azure/azure-event-hubs-java#287]
Looks to be related to #204, but we're seeing it in version 1.0.0.

## Versions  
- OS platform and version: Ubuntu
- Maven package version or commit ID: 1.0.0

error[UnHandled exception while processing events in reactor:
 java.lang.NullPointerException: uncorrelated channel: 0
 org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:112)
 org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
 org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:292)
 com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:445)
 java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
 java.util.concurrent.FutureTask.run(FutureTask.java:266)
 java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
 java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
 java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
 java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 java.lang.Thread.run(Thread.java:748)Cause: uncorrelated channel: 0
 org.apache.qpid.proton.engine.impl.TransportImpl.handleBegin(TransportImpl.java:1143)
 org.apache.qpid.proton.engine.impl.TransportImpl.handleBegin(TransportImpl.java:66)
 org.apache.qpid.proton.amqp.transport.Begin.invoke(Begin.java:144)
 org.apache.qpid.proton.engine.impl.TransportImpl.handleFrame(TransportImpl.java:1384)
 org.apache.qpid.proton.engine.impl.FrameParser.input(FrameParser.java:418)
 org.apache.qpid.proton.engine.impl.FrameParser.process(FrameParser.java:527)
 org.apache.qpid.proton.engine.impl.SaslImpl$SaslTransportWrapper.reallyProcessInput(SaslImpl.java:681)
 org.apache.qpid.proton.engine.impl.SaslImpl$SaslTransportWrapper.process(SaslImpl.java:630)
 org.apache.qpid.proton.engine.impl.HandshakeSniffingTransportWrapper.process(HandshakeSniffingTransportWrapper.java:101)
 org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.unwrapInput(SimpleSslTransportWrapper.java:135)
 org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.process(SimpleSslTransportWrapper.java:361)
 org.apache.qpid.proton.engine.impl.ssl.SslImpl$UnsecureClientAwareTransportWrapper.process(SslImpl.java:144)
 org.apache.qpid.proton.engine.impl.TransportImpl.process(TransportImpl.java:1495)
 org.apache.qpid.proton.reactor.impl.IOHandler$1.run(IOHandler.java:233)
 org.apache.qpid.proton.reactor.impl.SelectableImpl.readable(SelectableImpl.java:118)
 org.apache.qpid.proton.reactor.impl.IOHandler.handleQuiesced(IOHandler.java:60)
 org.apache.qpid.proton.reactor.impl.IOHandler.onUnhandled(IOHandler.java:389)
 org.apache.qpid.proton.engine.BaseHandler.onReactorQuiesced(BaseHandler.java:87)
 org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:206)
 org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
 org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
 org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:292)
 com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:445)
 java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
 java.util.concurrent.FutureTask.run(FutureTask.java:266)
 java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
 java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
 java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
 java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 java.lang.Thread.run(Thread.java:748)]

**Comment by SreeramGarlapati** [comment]
Hi @dgilling,
 Can you please explain - how I can reproduce - this error. or is it possible for you to enable INFO level tracing and provide logs for `com.microsoft.azure.eventhubs`.
 Thanks a lot for reporting the issue...

**Comment by SreeramGarlapati** [comment]
Hi @dgilling,

Thanks a lot for reporting this issue. I want to fix this issue - and have been trying to pinpoint the issue - but, cannot reproduce the error. Is it possible to enable INFO level tracing and forward logs to sreeramg@microsoft.com - when you encounter this error.

**Comment by SreeramGarlapati** [comment]
To update, what's going on with this issue:

1) This is a Reactor error. Impact-wise, when reactor error happens (i.e., when code called by Reactor Call-backs throws unexpected errors), Reactor stops and as a Result Event Hubs client also would stop. To prevent this - we already have recover mechanism - which restarts a new instance of reactor and copy the state from old to new reactor. But, this wasn't working due to the issue: #298. This was addressed.

2) Now, coming back to the current issue as such, this is an exception thrown by proton library - in the event, when the client receives a `BeginSession` frame from `Event Hubs service` - for which client did not initiate the call. Here's the [line no. in proton code for reference](https://github.com/apache/qpid-proton-j/blob/0.25.0/proton-j/src/main/java/org/apache/qpid/proton/engine/impl/TransportImpl.java#L1143). 

As far as the Event Hubs service `AmqpListener` implementation goes, this - is not possible, which is why, a timestamp of when this happened and the Namespace on which this error happened would really help us debug the issue. 

Appreciate your help @dgilling.

**Comment by vin-chi** [comment]
Hi Sreeram,

I got the same issue this is my exception trace

2018-07-05 17:05:59,831 AzureEventHubPartitionReader-0 (MessagingFactory.java:433) messagingFactory[MessagingFactory98faca], hostName[souvik-eventhub.servicebus.windows.net], info[starting reactor instance.]
2018-07-05 17:05:59,831 AzureEventHubPartitionReader-0 (ReactorHandler.java:32) reactor.onReactorInit
2018-07-05 17:06:01,806 AzureEventHubPartitionReader-0 (ConnectionHandler.java:131) onConnectionRemoteOpen: hostname[souvik-eventhub.servicebus.windows.net:5671, b0dfbdafc1df44c2bb60c94d0a130490_G2]
2018-07-05 17:06:01,822 AzureEventHubPartitionReader-2 (ReceiveLinkHandler.java:40) linkName[cbs:receiver], localSource[Source{address='$cbs', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, distributionMode=null, filter=null, defaultOutcome=null, outcomes=null, capabilities=null}]
2018-07-05 17:06:02,031 AzureEventHubPartitionReader-2 (SessionHandler.java:85) entityName[cbs-session], sessionIncCapacity[0], sessionOutgoingWindow[2147483647]
2018-07-05 17:06:02,731 AzureEventHubPartitionReader-0 (SendLinkHandler.java:38) linkName[cbs:sender], remoteTarget[Target{address='$cbs', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
2018-07-05 17:06:02,731 AzureEventHubPartitionReader-0 (SendLinkHandler.java:90) linkName[cbs:sender], unsettled[0], credit[100]
2018-07-05 17:06:02,732 AzureEventHubPartitionReader-0 (ReceiveLinkHandler.java:53) linkName[cbs:receiver], remoteSource[Source{address='$cbs', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, distributionMode=null, filter=null, defaultOutcome=null, outcomes=null, capabilities=null}]
2018-07-05 17:06:02,734 AzureEventHubPartitionReader-0 (SendLinkHandler.java:90) linkName[cbs:sender], unsettled[1], credit[99]
2018-07-05 17:06:03,358 AzureEventHubPartitionReader-0 (SessionHandler.java:85) entityName[perftestcachedthreadpool1/ConsumerGroups/StriimConsumerGroup/Partitions/0], sessionIncCapacity[0], sessionOutgoingWindow[2147483647]
2018-07-05 17:06:03,359 AzureEventHubPartitionReader-0 (PartitionReceiverImpl.java:257) receiverPath[RECEIVER IS NULL], action[createReceiveLink], offset[-1], sequenceNumber[null], enqueuedTime[null], inclusiveFlag[true]
2018-07-05 17:06:03,361 AzureEventHubPartitionReader-0 (ReceiveLinkHandler.java:40) linkName[1f05a7_0490_G2_1530790563361], localSource[Source{address='perftestcachedthreadpool1/ConsumerGroups/StriimConsumerGroup/Partitions/0', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, distributionMode=null, filter={apache.org:selector-filter:string=UnknownDescribedType{descriptor=apache.org:selector-filter:string, described=amqp.annotation.x-opt-offset >= '-1'}}, defaultOutcome=null, outcomes=null, capabilities=null}]
2018-07-05 17:06:03,671 AzureEventHubPartitionReader-0 (ReceiveLinkHandler.java:53) linkName[1f05a7_0490_G2_1530790563361], remoteSource[Source{address='perftestcachedthreadpool1/ConsumerGroups/StriimConsumerGroup/Partitions/0', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, distributionMode=null, filter={apache.org:selector-filter:string=org.apache.qpid.proton.codec.DecoderImpl$UnknownDescribedType@4248645c}, defaultOutcome=null, outcomes=null, capabilities=null}]
2018-07-05 17:06:03,672 AzureEventHubPartitionReader-0 (MessageReceiver.java:536) receiverPath[perftestcachedthreadpool1/ConsumerGroups/StriimConsumerGroup/Partitions/0], linkname[1f05a7_0490_G2_1530790563361], updated-link-credit[999], sentCredits[999], ThreadId[86]
2018-07-05 17:06:03,672 AzureEventHubPartitionReader-0 (MessageReceiver.java:307) receiverPath[perftestcachedthreadpool1/ConsumerGroups/StriimConsumerGroup/Partitions/0], linkname[1f05a7_0490_G2_1530790563361], updated-link-credit[999], sentCredits[999]
2018-07-05 17:06:03,674 AzureEventHubPartitionReader-3 (SendLinkHandler.java:90) linkName[cbs:sender], unsettled[1], credit[98]
2018-07-05 17:06:19,291 AzureEventHubPartitionReader-0 (SessionHandler.java:103) entityName[perftestcachedthreadpool1/ConsumerGroups/StriimConsumerGroup/Partitions/1], condition[Error{condition=null, description='null', info=null}]
com.microsoft.azure.eventhubs.EventHubException: session creation timedout., errorContext[NS: souvik-eventhub.servicebus.windows.net, PATH: perftestcachedthreadpool1/ConsumerGroups/StriimConsumerGroup/Partitions/1]2018-07-05 17:06:19,292 main (AzureEventHubReader.java:130) Problem initializing receivers com.microsoft.azure.eventhubs.EventHubException: session creation timedout., errorContext[NS: souvik-eventhub.servicebus.windows.net, PATH: perftestcachedthreadpool1/ConsumerGroups/StriimConsumerGroup/Partitions/1]

	at com.microsoft.azure.eventhubs.impl.SessionHandler$SessionTimeoutHandler.onEvent(SessionHandler.java:167)
	at com.microsoft.azure.eventhubs.impl.DispatchHandler.onTimerTask(DispatchHandler.java:9)
	at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:233)
	at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
	at com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:445)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
2018-07-05 17:06:23,534 AzureEventHubPartitionReader-3 (MessagingFactory.java:470) messagingFactory[MessagingFactory98faca], hostName[souvik-eventhub.servicebus.windows.net], error[UnHandled exception while processing events in reactor:
java.lang.NullPointerException: uncorrelated channel: 2
org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:112)
org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:292)
com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:445)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
java.lang.Thread.run(Thread.java:745)Cause: uncorrelated channel: 2
org.apache.qpid.proton.engine.impl.TransportImpl.handleBegin(TransportImpl.java:1143)
org.apache.qpid.proton.engine.impl.TransportImpl.handleBegin(TransportImpl.java:66)
org.apache.qpid.proton.amqp.transport.Begin.invoke(Begin.java:144)
org.apache.qpid.proton.engine.impl.TransportImpl.handleFrame(TransportImpl.java:1384)
org.apache.qpid.proton.engine.impl.FrameParser.input(FrameParser.java:418)
org.apache.qpid.proton.engine.impl.FrameParser.process(FrameParser.java:527)
org.apache.qpid.proton.engine.impl.SaslImpl$SaslTransportWrapper.reallyProcessInput(SaslImpl.java:681)
org.apache.qpid.proton.engine.impl.SaslImpl$SaslTransportWrapper.process(SaslImpl.java:630)
org.apache.qpid.proton.engine.impl.HandshakeSniffingTransportWrapper.process(HandshakeSniffingTransportWrapper.java:101)
org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.unwrapInput(SimpleSslTransportWrapper.java:135)
org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.process(SimpleSslTransportWrapper.java:361)
org.apache.qpid.proton.engine.impl.ssl.SslImpl$UnsecureClientAwareTransportWrapper.process(SslImpl.java:144)
org.apache.qpid.proton.engine.impl.TransportImpl.process(TransportImpl.java:1495)
org.apache.qpid.proton.reactor.impl.IOHandler$1.run(IOHandler.java:233)
org.apache.qpid.proton.reactor.impl.SelectableImpl.readable(SelectableImpl.java:118)
org.apache.qpid.proton.reactor.impl.IOHandler.handleQuiesced(IOHandler.java:60)
org.apache.qpid.proton.reactor.impl.IOHandler.onUnhandled(IOHandler.java:389)
org.apache.qpid.proton.engine.BaseHandler.onReactorQuiesced(BaseHandler.java:87)
org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:206)
org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:292)
com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:445)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
java.lang.Thread.run(Thread.java:745)]
2018-07-05 17:06:23,545 AzureEventHubPartitionReader-1 (MessagingFactory.java:433) messagingFactory[MessagingFactory98faca], hostName[souvik-eventhub.servicebus.windows.net], info[starting reactor instance.]
2018-07-05 17:06:23,546 AzureEventHubPartitionReader-1 (ReactorHandler.java:32) reactor.onReactorInit


The eventhub has 32 partitions and it has a lot of data around like 30 to 40 lakhs in each partition.

Code used to create receivers  (I first do a check to verify if the event hub exists and get the partition information from it.

`PartitionReceiver [] receivers;
  receivers = new PartitionReceiver[noOfPartitions];
            for(String i : pInfo.getPartitionIds()) {
                receivers[Integer.valueOf(i)] = client.createEpochReceiverSync(
                        EventHubClient.DEFAULT_CONSUMER_GROUP_NAME,
                        i.toString(),
                        eventPosition,
                        epoch++);   
            } `

What am I doing wrong ?

This is my connection string 
`connStr = new ConnectionStringBuilder()
        .setNamespaceName(eventhubNamespace.toString())
        .setEventHubName(eventhubName.toString())
        .setSasKeyName(sasKeyName.toString())
        .setSasKey(sasKey.toString())
        .setOperationTimeout(Duration.ofMinutes(3));`

**Comment by vin-chi** [comment]
What is the effective way of closing PartitionReceiver.
I have one PartitionReceiver per partition and closing via PartitionReceiver.closeSync(). The operation timeout set in connection string is 5 mins (I'm using version 1.0.1). And I always get

Close operation on Receive Link(07332a_c74_G26_1531559492424) timed out at 2018-07-14T02:14:13.391-07:00[PST8PDT], errorContext[NS: xxxx.servicebus.windows.net, PATH: test5mdsv4/ConsumerGroups/ehwritergroup/Partitions/0, REFERENCE_ID: 07332a_c74_G26_1531559492424, PREFETCH_COUNT: 999, LINK_CREDIT: 994, PREFETCH_Q_LEN: 4]

And on restart when i try to initialize the PartitionReceiver i get the following exception

com.microsoft.azure.eventhubs.EventHubException: session creation timedout., errorContext[NS: xxxx.servicebus.windows.net, PATH: ehteste1p1/ConsumerGroups/ehwritergroup/Partitions/1]
at com.microsoft.azure.eventhubs.impl.SessionHandler$SessionTimeoutHandler.onEvent(SessionHandler.java:159)
at com.microsoft.azure.eventhubs.impl.DispatchHandler.onTimerTask(DispatchHandler.java:9)
at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:233)
at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
at com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:462)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)

04:40:52,764WARN AzureEventHubWriterMDCheck-0 com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run:487 messagingFactory[MessagingFactoryab1d8d], hostName[striimeventhub.servicebus.windows.net], error[UnHandled exception while processing events in reactor:
java.lang.NullPointerException: uncorrelated channel: 3
org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:112)
org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:292)
com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:462)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
java.lang.Thread.run(Thread.java:745)Cause: uncorrelated channel: 3
org.apache.qpid.proton.engine.impl.TransportImpl.handleBegin(TransportImpl.java:1143)
org.apache.qpid.proton.engine.impl.TransportImpl.handleBegin(TransportImpl.java:66)
org.apache.qpid.proton.amqp.transport.Begin.invoke(Begin.java:144)
org.apache.qpid.proton.engine.impl.TransportImpl.handleFrame(TransportImpl.java:1384)
org.apache.qpid.proton.engine.impl.FrameParser.input(FrameParser.java:418)
org.apache.qpid.proton.engine.impl.FrameParser.process(FrameParser.java:527)
org.apache.qpid.proton.engine.impl.SaslImpl$SaslTransportWrapper.reallyProcessInput(SaslImpl.java:681)
org.apache.qpid.proton.engine.impl.SaslImpl$SaslTransportWrapper.process(SaslImpl.java:630)
org.apache.qpid.proton.engine.impl.HandshakeSniffingTransportWrapper.process(HandshakeSniffingTransportWrapper.java:101)
org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.unwrapInput(SimpleSslTransportWrapper.java:135)
org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper.process(SimpleSslTransportWrapper.java:361)
org.apache.qpid.proton.engine.impl.ssl.SslImpl$UnsecureClientAwareTransportWrapper.process(SslImpl.java:144)
org.apache.qpid.proton.engine.impl.TransportImpl.process(TransportImpl.java:1495)
org.apache.qpid.proton.reactor.impl.IOHandler$1.run(IOHandler.java:233)
org.apache.qpid.proton.reactor.impl.SelectableImpl.readable(SelectableImpl.java:118)
org.apache.qpid.proton.reactor.impl.IOHandler.handleQuiesced(IOHandler.java:60)
org.apache.qpid.proton.reactor.impl.IOHandler.onUnhandled(IOHandler.java:389)
org.apache.qpid.proton.engine.BaseHandler.onReactorQuiesced(BaseHandler.java:87)
org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:206)
org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:292)
com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:462)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
java.lang.Thread.run(Thread.java:745)]

**Comment by SreeramGarlapati** [comment]
Thanks a lot @vin-chi - this has been an issue for a while and I finally got a log - which can help reproduce this scenario. Owe you on that!

here's how to repro:
- configure the listener peer - to delay respond to session.begins by 5 secs
- then client (another peer) sends session.begin
- client doesn't receive any response even after 1 sec
- then the client decides to close the session
- then, after few sec's service(listener) responds back with session.begin

In this specific scenario - proton-j framework is not ready to handle this and throws a `NullPointerException`.

Proton-j folks have a todo to fix this - please file a JIRA issue on them to get this issue fixed:
https://github.com/apache/qpid-proton-j/blob/3e59d10d65f0206ddcd75be59749d28aab17402c/proton-j/src/main/java/org/apache/qpid/proton/engine/impl/TransportImpl.java#L1160

JavaClient will restart he reactor instance and recovers from this error.

**Comment by vin-chi** [comment]
@SreeramGarlapati - I don't see a Issues tab in [qpid-proton-j](https://github.com/apache/qpid-proton-j). Is there any other hack or fix that you can do from your end ?

One more info is the timeout exception on close in EventHubReader.close occurs only when a huge amount (like more than 10k per partition) of data is written to the eventhub (this happens with 2 or 30 partitions).  And reinitialising the receiver (with same consumer group) throws the exception specified in the previous comments.

**Comment by SreeramGarlapati** [comment]
This is where you can file issues for Qpid projects:
https://issues.apache.org/jira/secure/CreateIssue.jspa


Thanks for the input on Reader.close. Will look into this soon...

**Comment by vivianforzj** [comment]
@vin-chi hi, I also get the same exception. Do you have any idea to get around this problem? thx

**Comment by SreeramGarlapati** [comment]
@vin-chi - does this error bring down your process? This should be a recoverable error and you should not need to work around it. Please let us know - if that is not the case.

**Comment by balistof** [comment]
@SreeramGarlapati this does not exit the process (this would be actually a way I can deal with) but it simply blocks the thread which calls `EventHubClient.sendSync`.

**Comment by SreeramGarlapati** [comment]
@balistof - reactorInit log is spewed in our recovery logic. So, from that - `Reactor` instance is definitely recreated. As soon as it is created - it starts off create connection.
Is there any network error / outage going on during that time - which could result in a delay on connection open. Either ways, this should not have blocked the thread. Do you have any tracing that follows the reactor init trace - from the handler classes: `ConnectionHandler` etc. ?

**Comment by balistof** [comment]
Don't have a trace but it was sitting there forever until I killed the process. Freitag, 21 September 2018, 07:56nachm. +02:00 von Sreeram Garlapati  notifications@github.com :

**Comment by SreeramGarlapati** [comment]
I created this JIRA issue with PROTON folks: https://issues.apache.org/jira/browse/PROTON-1939 
- will keep this thread posted on further findings.

**Comment by SreeramGarlapati** [comment]
@balistof - here's an interim fix - before proton-j folks come back on the issue:
https://github.com/SreeramGarlapati/azure-event-hubs-java/commit/36e22a188adaeb3fa8a002f2da1c1289bd472f0d 
I am testing this.

**Comment by SreeramGarlapati** [comment]
Nevermind, after debugging through - I don't think this would work. Getting the `NPE` fixed from proton-j folks should be the right direction. Will work towards that.

**Comment by vin-chi** [comment]
> @vin-chi hi, I also get the same exception. Do you have any idea to get around this problem? thx

No I don't

**Comment by vin-chi** [comment]
> @vin-chi / @vivianforzj - does this error bring down your process? This should be a recoverable error and you should not need to work around it. Please let us know - if that is not the case.

Whenever the reader close throws a timeout exception, any other process or the same process is not able to initialize the PartitionReceiver to that EventHub. Say I have reader which was reading fine but i had to be stopped for some reason and while stopping the close throws an exception and that prevents any restart of the reader to the same eventhub, yes I'm blocked.

**Comment by SreeramGarlapati** [comment]
@vin-chi - thanks a lot for the response.
To understand the problem right, in your logs, i see 2 timestamps:
```
04:40:52,764WARN uncorrelated channel: 3

Close operation on Receive Link(07332a_c74_G26_1531559492424) timed out at 2018-07-14T02:14:13.391-07:00[PST8PDT], errorContext[NS: xxxx.servicebus.windows.net, PATH: 
```
which one happened first?

- did the `uncorrelated channel` error happen and then the subsequent receiveLink close calls are timing out ?

**Comment by SreeramGarlapati** [comment]
@balistof - if you are running into this problem during n/w errors and want to workaround it, heres one way:
change `com.microsoft.azure.eventhubs.impl.ClientConstants.SESSION_OPEN_TIMEOUT_IN_MS` to a higher value - 5 mins (basically, higher than tcp socket failure detection time). Ideal way would be to use `java reflection` and set this variable - on the app startup (or basically before any `EventHubClient` code usage).

However, please note that, this work around should be removed once `qpid-proton-j` library fixes the issue: https://issues.apache.org/jira/browse/PROTON-1939

### How does this resolve the issue:
EventHubs java client is designed on a `reactive-eventing` system. So, while reading through code-base, you may notice that, most of our layer involves `reactive-event-callbacks`. All runtime objects - of eventhubs - translates to - `amqp` entities. `Proton-j` library reads bytes off of TCP Sockets & translates them to `Amqp` artifacts and exposes them as `Amqp` events - through `Reactor` interface (`Basehandler` class has all events) - on which we react.

- `EventHubClient` maps to `AmqpConnection`.  
- `PartitionSender`, `PartitionReceiver` or the default sender instance created inside `EventHubClient` - translates to 1 `AmqpLink`. For each `AmqpLink` - we create 1 `AmqpSession`.
- All send & receive operations are mapped to `Amqp` link flow control.

In this case, SESSION_OPEN is a `Reactor` event and event hub library initiates this and waits for result. This constant -` SESSION_OPEN_TIMEOUT_IN_MS` - determines how long should the eventhubs library wait for `proton-j` to notify us of `session open result`. If, for some reason - this times out, we create new session and move on. And, this, in some cases, is resulting into an `NPE` in proton library. So, increasing it to a very large value will eliminate the possibility of hitting this. 

### Why does this timeout exist in the first place:
If user sets up TCP_SYN and ACK retries to a longer value - tcp socket failure detection time could be very high. If, in some of our customer environments, this is configured to higher values - higher than EventHubClient sdk timeouts, `Reactor` would not see any bytes on the transport (as the n/w layer is still retrying to connect), & wouldn't get back to us with `SESSION_OPEN`'s result for a very long time. As a result, there is no way to differentiate if some of the EventHubClient API is stuck or n/w is err'ing out. We added the SESSION_TIMEOUT as a way to ease debugging experience - during transient n/w errors.

**Comment by sjkwak** [comment]
This was fixed in the version 2.0 release.
